### PR TITLE
[CARBONDATA-3676] Support clean carbon data files of stages.

### PR DESCRIPTION
**Pull request by niuge01**
### Why is this PR needed?
 At the end of the CarbonInsertFromStageCommand, the stage files will be cleared, but the data files which referenced by stage files will be not cleared. This could lead to a large backlog of data files。
 
 ### What changes were proposed in this PR?
Provide a new command to allows us to delete data files which referenced by disabled table stages.

The new command is CarbonDeleteStageCommand.
    
 ### Does this PR introduce any user interface change?
 - Yes

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1845/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/141/

**Comment by niuge01** [comment]
retest this please

**Comment by QiangCai** [comment]
does this command support deleting any folder which specified?

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/143/

**Comment by niuge01** [comment]
> does this command support deleting any folder which specified?

Yes

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1847/

File: docs/dml-of-carbondata.md [hunk 1]
```diff
@@ -446,6 +446,46 @@ CarbonData DML statements are documented here,which includes:
   ```
   DELETE FROM carbontable WHERE column1 IN (SELECT column11 FROM sourceTable2 WHERE column1 = 'USA')
   ```
+    
+### DELETE STAGE
```
Comment (jackylk) [inline docs/dml-of-carbondata.md]:
```suggestion
### DELETE STAGE files
```

File: docs/dml-of-carbondata.md [hunk 2]
```diff
@@ -446,6 +446,46 @@ CarbonData DML statements are documented here,which includes:
...
+
+  This command allows us to delete data files which referenced by disabled table stages.
```
Comment (jackylk) [inline docs/dml-of-carbondata.md]:
```suggestion
  This command allows us to delete the data files (stage data) which is already loaded into the table.
```

File: docs/dml-of-carbondata.md [hunk 3]
```diff
@@ -446,6 +446,46 @@ CarbonData DML statements are documented here,which includes:
...
+  ```
+  DELETE FROM TABLE [db_name.]table_name STAGE OPTIONS(property_name=property_value, ...)
+  ```  
+  **Supported Properties:**
+
+| Property                                                     | Description                                                  |
+| ------------------------------------------------------------ | ------------------------------------------------------------ |
+| [data_file_location](#data_file_location)                    | The data files location                                      |
```
Comment (jackylk) [inline docs/dml-of-carbondata.md]:
```suggestion
| [location](#data_file_location)                    | The data file location                                      |
```

File: docs/dml-of-carbondata.md [hunk 4]
```diff
@@ -446,6 +446,46 @@ CarbonData DML statements are documented here,which includes:
...
+| [data_file_retain_time_second](#data_file_retain_time_second)| Data file retain time in second                              |
```
Comment (jackylk) [inline docs/dml-of-carbondata.md]:
```suggestion
| [retain_hour](#retain_hour)| Data file retain time in hours                              |
```

File: docs/dml-of-carbondata.md [hunk 5]
```diff
@@ -446,6 +446,46 @@ CarbonData DML statements are documented here,which includes:
...
+
+-
+  You can use the following options to delete data:
+
+  - ##### data_file_location: 
+    The data files location, the command will scan the location, and delete files which can be deleted.
```
Comment (jackylk) [inline docs/dml-of-carbondata.md]:
```suggestion
    The data files location, the command will scan the location, and delete files which are already loaded into the table. This check is done by scanning the stage metadata folder in table path.
```

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala [hunk 6]
```diff
@@ -218,6 +218,12 @@ abstract class CarbonDDLSqlParser extends AbstractCarbonSparkSQLParser {
         }
     }
 
+  protected lazy val options: Parser[(String, String)] =
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: you can rename `loadOptions` and reuse the same function

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+import org.apache.carbondata.core.statusmanager.StageInput
+import org.apache.carbondata.core.util.path.CarbonTablePath
+
+/**
+ * Delete carbon data files of table stages.
+ *
+ * @param databaseNameOp database name
+ * @param tableName      table name
+ */
+case class CarbonDeleteStageCommand(
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala]:
```suggestion
case class CarbonDeleteStageFilsCommand(
```

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+      }
+    ).filter(
+      stageDataFile => stageDataFile != null
+    ).map(
+      stageDataFile => Row.fromSeq(Seq(stageDataFile))
+    )
+    Seq.empty
+  }
+
+  private def listStageFiles(tablePath: String, configuration: Configuration): Seq[CarbonFile] = {
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala]: Is this for listing stage metadata files? It is not easy to tell the difference from `listStageDataFiles`

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    if (stageDirectory.exists()) {
+      stageDirectory.listFiles().filter { file =>
+        !file.getName.endsWith(CarbonTablePath.SUCCESS_FILE_SUBFIX)
+      }
+    } else {
+      Seq.empty
+    }
+  }
+
+  private def listStageDataFiles(location: String,
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala]: move `location` to next line

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    val stageDataFileLocation = FileFactory.getCarbonFile(location, configuration)
+    if (!stageDataFileLocation.exists()) {
+      LOGGER.warn("Stage data file location is not exists. " + location)
+      Seq.empty
+    } else {
+      stageDataFileLocation.listFiles(true).asScala
+    }
+  }
+
+  private def listStageDataFilesReferenced(stageFiles: Seq[CarbonFile],
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala]: move `stageFiles: Seq[CarbonFile]` to next line

**Comment by niuge01** [comment]
retest this please

File: docs/dml-of-carbondata.md (hunk 2 above)
Comment (niuge01) [reply docs/dml-of-carbondata.md]: Accept suggestion, and updated.

File: docs/dml-of-carbondata.md (hunk 4 above)
Comment (niuge01) [reply docs/dml-of-carbondata.md]: Accept suggestion, and updated.

File: docs/dml-of-carbondata.md (hunk 3 above)
Comment (niuge01) [reply docs/dml-of-carbondata.md]: Accept suggestion, and updated.

File: docs/dml-of-carbondata.md (hunk 5 above)
Comment (niuge01) [reply docs/dml-of-carbondata.md]: Accept suggestion, and updated.

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala (hunk 6 above)
Comment (niuge01) [reply integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: Done.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala (hunk 7 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala]: Accept suggestion, and updated.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala (hunk 8 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala]: The method name has changed to listStageMetadataFiles

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala (hunk 9 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala]: Done.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala (hunk 10 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageCommand.scala]: Done.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/153/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1856/

**Comment by jackylk** [comment]
LGTM

**Comment by QiangCai** [comment]
Better to avoid to specify the path if the path can be any path.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+          null
+        } else {
+          // Stage data file deleted.
+          stageDataFile.getCanonicalPath
+        }
+      }
+    ).filter(
+      stageDataFile => stageDataFile != null
+    ).map(
+      stageDataFile => Row.fromSeq(Seq(stageDataFile))
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala]: replace with .collect()
No need to loop twice

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+ * @param databaseNameOp database name
+ * @param tableName      table name
+ */
+case class CarbonDeleteStageFilesCommand(
+    databaseNameOp: Option[String],
+    tableName: String,
+    options: Map[String, String]
+) extends DataCommand {
+
+  @transient var LOGGER: Logger = _
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala]: change to val LOGGER = LogServiceFactory.getLogService(this.getClass.getCanonicalName)

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+
+  override def processData(spark: SparkSession): Seq[Row] = {
+    LOGGER = LogServiceFactory.getLogService(this.getClass.getCanonicalName)
+    Checker.validateTableExists(databaseNameOp, tableName, spark)
+    val table = CarbonEnv.getCarbonTable(databaseNameOp, tableName)(spark)
+    val configuration = spark.sessionState.newHadoopConf()
+    FileFactory.getConfiguration.addResource(configuration)
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala]: Why is this required?

**Comment by kunal642** [comment]
@niuge01 Why are you not using the executors to make this job parallel? 
Not recommended to open threads in the driver, it will impact the other queries

File: docs/dml-of-carbondata.md (hunk 1 above)
Comment (Indhumathi27) [inline docs/dml-of-carbondata.md]: Can you please add documentation for flink integration to carbon and can move these new DDL's added for flink and stage to it.

Comment (niuge01) [reply docs/dml-of-carbondata.md]: No DDL's added  for flink.
Both CarbonInsertFromStageCommand and CarbonDeleteStageFilesCommand are DML.

Comment (Indhumathi27) [reply docs/dml-of-carbondata.md]: okay.. Can you please add documentation for flink integration to carbon and stage files

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala (hunk 11 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala]: Done.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala (hunk 12 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala]: Done

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala (hunk 13 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala]: Removed

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/166/

**Comment by niuge01** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/169/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1869/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1871/

File: integration/flink/src/test/scala/org/apache/carbon/flink/TestDeleteStageFiles.scala [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+class TestDeleteStageFiles extends QueryTest {
+
+  val tableName = "test_flink_delete_stage_file"
+
+  test("Delete stage file success") {
+    sql(s"DROP TABLE IF EXISTS $tableName").collect()
+    sql(
+      s"""
+         | CREATE TABLE $tableName (stringField string, intField int, shortField short)
```
Comment (jackylk) [inline integration/flink/src/test/scala/org/apache/carbon/flink/TestDeleteStageFiles.scala]: How about partition table, is there testcase?

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonS3Property.java [hunk 15]
```diff
@@ -27,6 +27,7 @@
 
   static final String DATA_TEMP_PATH = "carbon.writer.s3.data.temp.path";
 
+  @Deprecated
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonS3Property.java]: I think we can delete this constant directly, since this version has not been released

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+      stageDirectory.listFiles().filter { file =>
+        !file.getName.endsWith(CarbonTablePath.SUCCESS_FILE_SUBFIX)
+      }
+    } else {
+      Seq.empty
+    }
+  }
+
+  private def listStageDataFiles(
+      location: String,
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala]: can you pass tablePath instead, make it same as listStageMetadataFiles

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+    val stageDataFileLocation = FileFactory.getCarbonFile(location, configuration)
+    if (!stageDataFileLocation.exists()) {
+      LOGGER.warn("Stage data file location is not exists. " + location)
+      Seq.empty
+    } else {
+      stageDataFileLocation.listFiles(true).asScala
+    }
+  }
+
+  private def listStageDataFilesReferenced(
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala]: can you add comment to describe the functionality of this function

**Comment by niuge01** [comment]
retest this please

File: docs/dml-of-carbondata.md (hunk 1 above)
Comment (niuge01) [reply docs/dml-of-carbondata.md]: OK

File: integration/flink/src/test/scala/org/apache/carbon/flink/TestDeleteStageFiles.scala (hunk 14 above)
Comment (niuge01) [reply integration/flink/src/test/scala/org/apache/carbon/flink/TestDeleteStageFiles.scala]: added

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonS3Property.java (hunk 15 above)
Comment (niuge01) [reply integration/flink/src/main/java/org/apache/carbon/flink/CarbonS3Property.java]: deleted.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala (hunk 16 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala]: done.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala (hunk 17 above)
Comment (niuge01) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonDeleteStageFilesCommand.scala]: added

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/173/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1876/
